### PR TITLE
Memory fix

### PR DESCRIPTION
**Pull request by michel2323**
We finally found it! Here's the fix.

### CODE REVIEW

**Comment by cnpetra** [comment]
I don't see how this would have fixed the memory issue... could you please elaborate how that piece of code was causing that many calls to StochVector constructor?

**Comment by michel2323** [comment]
`makeVariables` is called in a lot places. Each time `newPrimal` etc were called. This called the `Empty` variants instead if `nx`, `my` or `mz` is 0, respectively.

**Comment by michel2323** [comment]
Cosmin, like always I pushed with a bug. However, I now know exactly where the memory usage comes from. I restructured a bit such that the `Empty()` functions for `newPrimalVector` etc are called whenever possible. Now it's clear that all the allocations come from the `StochDummyVector`s in the these `Empty()` calls. The dummy just takes too much memory. I'm thinking on how to resolve this gracefully. Any ideas?

**Comment by michel2323** [comment]
This saves now considerably memory going from 1.8 GB to 600 GB for 100,000 scenarios. I will probably also change to a `map` or something for `children` `std::vector`s. By overloading `push_back` and `[]` I can skip the `push_back` of the dummies and just return the dummy for the `[]` operator if it accesses an empty entry. Maybe I won't even need map.

**Comment by michel2323** [comment]
@cnpetra Everywhere the tree has been replicated for other datastructures (`sVars, sData, some matrices, `sLinsys`) I replaced the `vector<>children` with a `map`. The dummies are not stored in map. This 'compression' is handled by the class `VectorCompressedDummy` in `StochVector.h`. 

This reduces the memory consumption of 100,000 scenarios from 1.8 GB to 150 MB. The reason for this leftover is that the tree `sTree` itself is still stored fully as a vector, with the nonlocal children marked with `MPI_COMM_NULL`. I did not do the trick here, because this is linked to the partitioning in `assignProcess()`. This can be done in the future, but I don't want to touch partitioning now.

File: PIPS-NLP/Core/StochLinearAlgebra/StochVector.h [hunk 1]
```diff
@@ -11,10 +11,53 @@
...
+   if (child != sTree::dummy) {
+     children[size_] = child;
+   }
+   size_++;
+ }
+
+private:
+  std::unordered_map<int, pTree> children;
+  size_t size_ = 0;
+  size_t tablesize_ = 0;
```
Comment (cnpetra) [inline PIPS-NLP/Core/StochLinearAlgebra/StochVector.h]: @michel2323  is "tablesize_" used anywhere?

**Comment by cnpetra** [comment]
> @cnpetra This saves now considerably memory going from 1.8 GB to 600 GB for 100,000 scenarios. I will probably also change to a `map` or something for `children` `std::vector`s. By overloading `push_back` and `[]` I can skip the `push_back` of the dummies and just return the dummy for the `[]` operator if it accesses an empty entry. Maybe I won't even need `map`.

you should me give me some credit though for the idea ;)

**Comment by michel2323** [comment]
> > @cnpetra This saves now considerably memory going from 1.8 GB to 600 GB for 100,000 scenarios. I will probably also change to a `map` or something for `children` `std::vector`s. By overloading `push_back` and `[]` I can skip the `push_back` of the dummies and just return the dummy for the `[]` operator if it accesses an empty entry. Maybe I won't even need `map`.
> 
> you should me give me some credit though for the idea ;)

If you look at the timestamp, this was 15min before your email ;-). But in any case. You helped a whole lot! Wouldn't have boiled down the problem. I used a few too many core hours on Theta :-/.

File: PIPS-NLP/Core/StochLinearAlgebra/StochVector.h (hunk 1 above)
Comment (michel2323) [reply PIPS-NLP/Core/StochLinearAlgebra/StochVector.h]: fixed!

**Comment by michel2323** [comment]
I tested on Theta and it works fine. So it should now be ready to be merged.

**Comment by cnpetra** [comment]
> > > @cnpetra This saves now considerably memory going from 1.8 GB to 600 GB for 100,000 scenarios. I will probably also change to a `map` or something for `children` `std::vector`s. By overloading `push_back` and `[]` I can skip the `push_back` of the dummies and just return the dummy for the `[]` operator if it accesses an empty entry. Maybe I won't even need `map`.
> > 
> > 
> > you should me give me some credit though for the idea ;)
> 
> If you look at the timestamp, this was 15min before your email ;-). But in any case. You helped a whole lot! Wouldn't have boiled down the problem. I used a few too many core hours on Theta :-/.
> 
> Two fools, same idea :-D.

:1st_place_medal: for you

**Pull request merged**